### ISSUE TITLE
Fixing random seeds does not yield deterministic results

### ISSUE THREAD

**Issue by mufeili** [dmlc/dgl#412]
- [Related discussion thread](https://discuss.dgl.ai/t/how-to-fix-random-seeds-in-dgl-in-training/65/3)

It has been observed by Erutan-pku, @hbsun2113 , me and @yzh119 that after fixing NumPy random seed, PyTorch random seed and `cudnn.deterministic=True` we still get different results across runs. There can be two possibilities:
1. This is a bug.
2. Some system optimization involves randomness, in which case we should provide APIs for users to fix results.

**Comment by BarclayII** [comment]
In the case of SGC, the function `python/dgl/data/citation_graph.py:_encode_onehot` does *not* necessarily return the same output from the same input, since it is `enumerat`ing a `set`.

Changing from `classes = set(labels)` to `classes = list(sorted(set(labels)))` solves the problem.

**Comment by mufeili** [comment]
@BarclayII In that case we may want to add an option for our API to decide if a deterministic split is to be employed. Could you also please check GraphSAGE as mentioned by @hbsun2113 ?

**Comment by BarclayII** [comment]
Works for me.

**Comment by BarclayII** [comment]
> @BarclayII In that case we may want to add an option for our API to decide if a deterministic split is to be employed. Could you also please check GraphSAGE as mentioned by @hbsun2113 ?

Also it's not related to splitting the dataset, but encoding the labels into one-hot vectors.

**Comment by hbsun2113** [comment]
> Changing from `classes = set(labels)` to `classes = list(sorted(set(labels)))` solves the problem.

I wonder what is the difference between of them?

**Comment by mufeili** [comment]
@hbsun2113 Sets are unordered and you can get different results across tries. To sort it can solve the issue.

**Comment by hbsun2113** [comment]
> Sets are unordered and you can get different results across tries. To sort it can solve the issue.

@mufeili  Encoding label to different onehot leads to different results?

**Comment by BarclayII** [comment]
> @mufeili Encoding label to different onehot leads to different results?
> I think we only need call `_encode_onehot` once during training.

Yes to both.

Think of a linear logistic regression.  If we swap the positive and negative labels, the optimal parameters would switch their signs as well.  This implies that, given everything the same (including initial parameters) except that we permute the labels, the trajectory of parameter updates would still be different.

Now that we have a complicated non-convex model, we know that a lot of local minima exist.  Since the trajectory of parameter updates differ if we permute the labels, it is natural that the model performance would be (slightly) different between runs.  This also naturally extends to multi-class classification.

**Comment by michaeljyt** [comment]
I just find an issue worth noting. When I fix all random seeds (torch, cuda, numpy, random,python), I can get same results using CPU but different results with GPU. 
Is there any function in DGL using GPU atomic function? This would introduce troubles in reproducity. Thanks.

**Comment by VoVAllen** [comment]
@michaeljyt Currently all operation in dgl are using framework operator, which you can find [here](https://github.com/dmlc/dgl/blob/master/python/dgl/backend/pytorch/tensor.py). Using reduce function `fn.sum` will call `spmm` function and it might be non-deterministic. We will introduce new kernel in 0.3, using cuSparse kernel which is deterministic. (Correct me if I'm wrong @jermainewang @liangfu )

**Comment by michaeljyt** [comment]
@VoVAllen Thank very much for your reply. In my experiment, non-deterministic operation affects the performance seriously. CuSparse is a good choice. From the cuda document, the CUSPARSE_COOMM_ALG2 for cusparseSpMM is deterministic. Hope the new version would come out soon.

**Comment by michaeljyt** [comment]
> @michaeljyt Currently all operation in dgl are using framework operator, which you can find [here](https://github.com/dmlc/dgl/blob/master/python/dgl/backend/pytorch/tensor.py). Using reduce function `fn.sum` will call `spmm` function and it might be non-deterministic. We will introduce new kernel in 0.3, using cuSparse kernel which is deterministic. (Correct me if I'm wrong @jermainewang @ylfdq1118 )

Hi VoVAllen, I just tested torch.sparse function and find it quite slow (even worse than cpu). I am not familiar with CuSparse, so you may please test its speed first.

**Comment by VoVAllen** [comment]
@michaeljyt Another option is to use `scatter_add` to do the neighborhood sum at current stage, which is deterministic and had better performance.

**Comment by michaeljyt** [comment]
> @michaeljyt Another option is to use `scatter_add` to do the neighborhood sum at current stage, which is deterministic and had better performance.

Unfortunately, scatter_add is not deterministic using gpu... The document of Pytorch has clarified it... Anyway, if dgl is used in supervised learning, this issue is not important. :D
